sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -164,146 +152,121 @@ public void forceFlush() {
+
+    @Nullable private volatile CompletableResultCode flushRequested = null;
+    private volatile boolean continueWork = true;
+    private final ArrayList<SpanData> batch;
 
     private Worker(
         SpanExporter spanExporter,
         long scheduleDelayMillis,
-        int maxQueueSize,
         int maxExportBatchSize,
-        int exporterTimeoutMillis) {
+        int exporterTimeoutMillis,
+        BlockingQueue<ReadableSpan> queue) {
       this.spanExporter = spanExporter;
-      this.scheduleDelayMillis = scheduleDelayMillis;
-      this.maxQueueSize = maxQueueSize;
-      this.halfMaxQueueSize = maxQueueSize >> 1;
+      this.scheduleDelayNanos = TimeUnit.MILLISECONDS.toNanos(scheduleDelayMillis);
       this.maxExportBatchSize = maxExportBatchSize;
       this.exporterTimeoutMillis = exporterTimeoutMillis;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.queue = queue;
+      this.batch = new ArrayList<>(this.maxExportBatchSize);
     }
 
     private void addSpan(ReadableSpan span) {
-      synchronized (monitor) {
-        if (spansList.size() == maxQueueSize) {
-          droppedSpans.add(1);
-          return;
-        }
-        // TODO: Record a gauge for referenced spans.
-        spansList.add(span);
-        // Notify the worker thread that at half of the queue is available. It will take
-        // time anyway for the thread to wake up.
-        if (spansList.size() >= halfMaxQueueSize) {
-          monitor.notifyAll();
-        }
+      if (!queue.offer(span)) {
+        droppedSpans.add(1);
       }
     }
 
     @Override
     public void run() {
-      while (!Thread.currentThread().isInterrupted()) {
-        // Copy all the batched spans in a separate list to release the monitor lock asap to
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+      updateNextExportTime();
+
+      while (continueWork) {
+        try {
+          if (flushRequested != null) {
+            flush();
+          }
+
+          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
+          if (lastElement != null) {
+            batch.add(lastElement.toSpanData());
+          }
```

Comment:
It's probably a nice win to add a `drainTo` here to get as many elements as possible. In the case where there are constantly spans, we'll be able to handle them with less queue lock acquisitions and tighter loop. `drainTo` has this very tight loop inside the lock which JVM should deal well with.

```java
while (i < n) {
                    @SuppressWarnings("unchecked")
                    E e = (E) items[take];
                    c.add(e);
                    items[take] = null;
                    if (++take == items.length) take = 0;
                    i++;
                }
```

Reply:
A, main reason I cannot use `drain` here is that I immediately convert to `SpanData`. And I don't want to drain to one collection, then iterate and convert element by element into another collection.